x/marker/keeper/send_restrictions.go

```diff
@@ -0,0 +1,140 @@
+	if hasAccess || marker.AddressHasAccess(caller, types.Access_Transfer) || caller.Equals(k.markerModuleAddr) {
+		return nil
+	}
+
+	if len(marker.GetRequiredAttributes()) == 0 {
+		return fmt.Errorf("%s does not have transfer permissions", caller.String())
+	}
+	contains, err := k.ContainsRequiredAttributes(ctx, marker.GetRequiredAttributes(), to)
+	if err != nil {
+		return err
+	}
+	if !contains {
+		return fmt.Errorf("address %s does not contain the required attributes %v", to, marker.GetRequiredAttributes())
+	}
+	return nil
+}
+
+// NormalizeRequiredAttributes normalizes the required attribute names using name module's Normalize method
+func (k Keeper) NormalizeRequiredAttributes(ctx sdk.Context, requiredAttributes []string) ([]string, error) {
+	maxLength := int(k.attrKeeper.GetMaxValueLength(ctx))
+	result := make([]string, len(requiredAttributes))
+	for i, attr := range requiredAttributes {
+		if len(attr) > maxLength {
+			return nil, fmt.Errorf("required attribute %v length is too long %v : %v ", attr, len(attr), maxLength)
+		}
+
+		// for now just check if required attribute starts with a *
+		var prefix string
+		if ContainsWildCard(attr) {
+			prefix = attr[:2]
+			attr = attr[2:]
+		}
+		normalizedAttr, err := k.nameKeeper.Normalize(ctx, attr)
+		if err != nil {
+			return nil, err
+		}
+		result[i] = fmt.Sprintf("%s%s", prefix, normalizedAttr)
+	}
+	return result, nil
+}
+
+// ContainsRequiredAttributes retrieves the attributes from address and checks that all required attributes are present
+func (k Keeper) ContainsRequiredAttributes(ctx sdk.Context, requiredAttributes []string, address string) (bool, error) {
+	attributes, err := k.attrKeeper.GetAllAttributes(ctx, address)
+	if err != nil {
+		return false, err
+	}
+	return EnsureAllRequiredAttributesExist(requiredAttributes, attributes), nil
+}
+
+// EnsureAllRequiredAttributesExist checks that all requiredAttributes are in attributes list
+func EnsureAllRequiredAttributesExist(requiredAttributes []string, attributes []attrTypes.Attribute) bool {
+	for _, reqAttr := range requiredAttributes {
+		var match bool
+		for _, attr := range attributes {
+			match = MatchAttribute(reqAttr, attr.Name)
+			if match {
+				break
+			}
+		}
+		if !match {
+			return false
+		}
+	}
+	return true
+}
+
+// GetAddressHasAccess returns the bool value of address has access defaults to false if not set
+func GetAddressHasAccess(ctx sdk.Context) (bool, error) {
+	hasAccess := ctx.Value(AddressHasAccessKey)
+	if hasAccess == nil {
+		return false, nil
+	}
+	accessAllowed, success := hasAccess.(bool)
+	if !success {
+		return false, fmt.Errorf("incorrect type for context %s value", AddressHasAccessKey)
+	}
+	return accessAllowed, nil
+}
+
+// WithAddressHasAccess returns the context with the address has access set used for allowing address to send restricted markers
+func WithAddressHasAccess(ctx sdk.Context, hasAccess bool) sdk.Context {
```

Comment:
I kind of expected this to be named something more like `WithMarkerSendRestrictionBypass`. It's a little confusing to me to call it "address has access" when there's multiple addresses involved an this doesn't identify which one.

I'd also change `GetAddressHasAccess` to `HasMarkerSendRestrictionBypass` and have it only return a bool and return false if there's any error, basically `return success && accessAllowed`.

This can also be called first thing in `AllowMarkerSend` to skip over some lookups and processing.